book-content/chapters/09-typescript-only-features.md

```diff
@@ -0,0 +1,518 @@
+Based on what I've told you so far, you might be thinking of TypeScript as just "JavaScript with types". JavaScript handles the runtime code, and TypeScript describes it with types.
+
+But TypeScript actually has a few runtime features that don't exist in JavaScript. These features are compiled into JavaScript, but they are not part of the JavaScript language itself.
+
+In this chapter we'll look at several of these TypeScript-only features, including parameter properties, enums and namespaces. Along the way, we'll discuss benefits and trade-offs, as well as when you might want to stick with JavaScript.
+
+## Class Parameter Properties
+
+One TypeScript feature that doesn't exist in JavaScript is class parameter properties. These allow you to declare and initialize class members directly from the constructor parameters.
+
+Consider this `Rating` class:
+
+```typescript
+class Rating {
+  constructor(public value: number, private max: number) {}
+}
+```
+
+That the constructor includes `public` before the `value` parameter and `private` before the `max` parameter. In JavaScript, this compiles down to code which assigns the parameters to properties on the class:
+
+```typescript
+class Rating {
+  constructor(value, max) {
+    this.value = value;
+    this.max = max;
+  }
+}
+```
+
+Compared to handling the assignment manually, this saves a lot of code and keeps the class definition concise.
+
+But unlike other TypeScript features, the outputted JavaScript is not a direct representation of the TypeScript code. This can make it difficult to understand what's happening if you're not familiar with the feature.
```

Comment:
It can also introduce (at least at the moment) some unexpected issues in your code. You have mentioned before that class properties are initialized before the constructor gets executed. This is a fair approximation. So what about those properties? We could say the same about them - they are initialized at the beginning of the constructor. So which one runs first? Well, take a look at this:
```ts
// input
class Foo {
  prop = "foo";
  constructor(public foo: string = this.prop) {}
}

// output
"use strict";
class Foo {
    constructor(foo = this.prop) {
        this.foo = foo;
        this.prop = "foo";
    }
}
```

Oops, not cool 😉